frontend/src/main/scala/bloop/engine/BuildLoader.scala

```diff
@@ -22,56 +27,136 @@ object BuildLoader {
+              }
+            }
+        }
+      }
+    }
 
-    val all = configFiles.map(f => Task(Project.fromBytesAndOrigin(f.bytes, f.origin, logger)))
-    val groupTasks = all.grouped(10).map(group => Task.gatherUnordered(group)).toList
-    Task.sequence(groupTasks).map(_.flatten).executeOn(ExecutionContext.ioScheduler)
+    incrementalLoadTask.flatten.executeOn(ExecutionContext.ioScheduler)
   }
 
   /**
-   * Load all the projects from `configDir` in a parallel, lazy fashion via monix Task.
+   * Resolves the semanticdb plugin for every project with a different Scala
+   * version and then enables the Metals settings on those projects that have
+   * changed in this incremental build load.
    *
-   * @param configDir The base directory from which to load the projects.
-   * @param logger The logger that collects messages about project loading.
-   * @return The list of loaded projects.
+   * Some of the logic of this method has a duplicate version in
+   * `loadSynchronously` that doesn't use `Task` to resolve SemanticDB plugins
+   * in parallel and apply the Metals projects. Any change here **must** also
+   * be propagated there.
    */
-  def load(
+  private def resolveSemanticDBForProjects(
+      rawProjects: List[Project],
       configDir: AbsolutePath,
+      semanticDBVersion: String,
+      supportedScalaVersions: List[String],
       logger: Logger
-  ): Task[List[Project]] = {
-    val configFiles = readConfigurationFilesInBase(configDir, logger).map { ap =>
-      Task {
-        val bytes = ap.path.readAllBytes
-        val hash = ByteHasher.hashBytes(bytes)
-        Build.ReadConfiguration(Origin(ap, hash), bytes)
+  ): Task[List[(Project, Option[Project])]] = {
+    val projectsWithNoScalaConfig = new mutable.ListBuffer[Project]()
+    val groupedProjectsPerScalaVersion = new mutable.HashMap[String, List[Project]]()
+    rawProjects.foreach { project =>
+      project.scalaInstance match {
+        case None => projectsWithNoScalaConfig.+=(project)
+        case Some(instance) =>
+          val scalaVersion = instance.version
+          groupedProjectsPerScalaVersion.get(scalaVersion) match {
+            case Some(projects) =>
+              groupedProjectsPerScalaVersion.put(scalaVersion, project :: projects)
+            case None => groupedProjectsPerScalaVersion.put(scalaVersion, project :: Nil)
+          }
       }
     }
 
-    Task
-      .gatherUnordered(configFiles)
-      .flatMap(fs => loadBuildFromConfigurationFiles(configDir, fs, logger))
+    val enableMetalsInProjectsTask = groupedProjectsPerScalaVersion.toList.map {
+      case (scalaVersion, projects) =>
+        def enableMetalsTask(plugin: Option[AbsolutePath]) = {
+          Task(
+            projects.map(p => Project.enableMetalsSettings(p, configDir, plugin, logger) -> Some(p))
+          )
+        }
+
+        // Recognize 2.12.8-abdcddd as supported if 2.12.8 exists in supported versions
```

Comment:
Trick so that people using custom versions of Scala locally can have this working. This can happen sometimes to me :)